File: python/bowmark_web/_guard.py
"""The argument guard — refuse a value the wire cannot carry BEFORE the request.

── This is a RESTATEMENT of ``node/src/guard.ts``, not a translation ────────

The TypeScript guard is a checked, byte-for-byte copy of ``wireProblem`` in
``packages/schema/src/wire.ts``, and a test asserts identical ``{path, reason}``
strings for a 32-entry fixture table. **That is deliberately NOT the contract here**,
and the difference is a fact about the languages rather than a gap:

* The two runtimes refuse different THINGS. JavaScript has ``bigint``, ``Symbol``,
  ``Map`` and ``NaN``; Python has ``set``, ``bytes``, ``datetime``, ``Decimal`` and a
  ``float('nan')``. Neither list is a subset of the other, so a shared table of
  refusal reasons would have to describe values that cannot exist on one side.
* A refusal MESSAGE names the offending Python type, because that is what the caller
  is looking at. Telling a Python programmer their value "is a Map instance" would be
  worse than saying nothing.

What IS shared, and what ``tests/test_guard.py`` asserts against a table mirroring the
TypeScript one case for case, is the RULE: only JSON crosses — ``dict`` with string
keys, ``list``, ``str``, finite ``float``/``int``, ``bool``, ``None`` — everything
else is refused with its path, and a circular structure is refused rather than
overflowing the stack.

Why a walk and not ``json.dumps``: ``dumps`` accepts things that then arrive as
something else. A ``tuple`` becomes an array, an ``int`` key becomes a string key, a
subclass of ``dict`` serializes as its base, and ``allow_nan=True`` (the default)
emits bare ``NaN``, which is not JSON and which our API rejects at the parser. A
try/except around serialization therefore catches almost nothing, which is exactly the
class of silent lossiness Temporal shipped, diagnosed as a typing problem, and closed
won't-fix.
"""

from __future__ import annotations

import math
from typing import Any, NamedTuple, Sequence


class WireProblem(NamedTuple):
    """What is wrong with a value, and WHERE.

    The path is Playwright's format — ``a.inner[1].property`` — because naming the
    offending path is the difference between a caller fixing the bug in a minute and
    bisecting their own arguments.
    """

    path: str
    reason: str


def wire_problem(value: Any) -> WireProblem | None:
    """Walk ``value`` and return the FIRST thing the wire refuses, or ``None``."""
    return _walk(value, "", set())


def _walk(value: Any, path: str, seen: set[int]) -> WireProblem | None:
    if value is None:
        return None
    # `bool` before `int`, because `isinstance(True, int)` is True in Python and the
    # order is the only thing that stops a bool being reported as a number.
    if isinstance(value, bool) or isinstance(value, str):
        return None
    if isinstance(value, int):
        # Python's int is arbitrary-precision. JSON has no such thing, and a value
        # past IEEE-754's exact-integer range silently loses digits at whatever
        # parses it on the other side — which is the same silent-lossy class as
        # JavaScript's bigint, arriving without a distinct type to catch it on.
        if abs(value) > 2**53 - 1:
            return WireProblem(
                path,
                f"the integer {value} is past 2**53-1, so JSON cannot carry it exactly",
            )
        return None
    if isinstance(value, float):
        # `json.dumps` emits bare `NaN`/`Infinity` by default. Those are not JSON,
        # and this is the silent-lossy case the guard exists to catch.
        if not math.isfinite(value):
            return WireProblem(path, f"the number {value} has no JSON form")
        return None

    identity = id(value)
    if identity in seen:
        return WireProblem(path, "a circular reference")
    seen.add(identity)
    try:
        if isinstance(value, list):
            for index, item in enumerate(value):
                problem = _walk(item, f"{path}[{index}]", seen)
                if problem is not None:
                    return problem
            return None

        # A NOMINAL refusal, and the only one available at runtime. A `tuple`, a
        # `set`, a `datetime`, a `Decimal` and any class instance are all refused
        # here rather than being quietly flattened into an array, a string, or
        # whatever `default=` would have made of them.
        #
        # `type(value) is dict` rather than `isinstance`: a `dict` SUBCLASS
        # (`defaultdict`, `Counter`, a pydantic model's `__dict__`-alike) serializes
        # as its base and loses whatever made it a subclass, which is exactly the
        # lossiness this guard refuses elsewhere.
        if type(value) is not dict:
            return WireProblem(
                path,
                f"a {type(value).__name__} — only dicts, lists, strings, finite numbers, "
                "bools and None cross",
            )

        for key, child in value.items():
            if not isinstance(key, str):
                return WireProblem(
                    path,
                    f"an object with a {type(key).__name__} key ({key!r}) — JSON keys are strings, "
                    "and json.dumps would silently stringify it",
                )
            child_path = f"{path}.{key}" if path else key
            problem = _walk(child, child_path, seen)
            if problem is not None:
                return problem
        return None
    finally:
        seen.discard(identity)


def assert_wire_safe_args(label: str, args: Sequence[Any]) -> None:
    """Refuse the whole argument list before a byte leaves the process.

    Raises ``BowmarkError`` with code ``wire_refused``, naming the exact position:
    ``bowmark.hotels.search(args[0].checkIn)``. Failure is loud here for the reason EF
    Core made silent client-side fallback an error: a boundary that quietly accepts
    something it will mangle produces behaviour that only breaks in production.
    """
    from ._transport import BowmarkError

    for index, arg in enumerate(args):
        problem = wire_problem(arg)
        if problem is None:
            continue
        if not problem.path:
            where = f"args[{index}]"
        elif problem.path.startswith("["):
            where = f"args[{index}]{problem.path}"
        else:
            where = f"args[{index}].{problem.path}"
        raise BowmarkError(
            f"{label} was not called: {where} is {problem.reason}. "
            "Only JSON — dicts, lists, strings, finite numbers, bools and None — crosses "
            "to a capability.",
            code="wire_refused",
            path=label,
        )

File: node/src/guard.ts
// The argument guard — refuse a value the wire cannot carry BEFORE the request.
//
// ── This is a deliberate COPY of `wireProblem` in `packages/schema/src/wire.ts` ──
//
// Not an import, and not an oversight. `@bowmark/schema` is a workspace package;
// importing it would put `"@bowmark/schema": "workspace:*"` in the published
// `package.json` and the tarball would not install for anyone outside this repo.
// The package's zero-dependency promise is the reason this file exists at all.
//
// A copy drifts, so the copy is CHECKED: `tests/unit/bowmark-web-guard.test.ts`
// runs both implementations over one fixture table and asserts identical
// `{ path, reason }` for every entry. A new refusal in `wire.ts` that is not
// mirrored here fails that test by name.
//
// Why a walk and not a `try { JSON.stringify(args) }`: stringify throws on exactly
// two things, a circular structure and a `BigInt`. Everything else it "handles"
// lossily — `Date` → string, `Map`/`Set` → `{}`, class instance → plain object,
// function-valued key → dropped — so a try/catch catches almost nothing. Temporal
// shipped that exact bug, diagnosed it as a typing problem, and closed it won't-fix.

import { VALIDATORS } from "./generated/validators.js";
import { BowmarkError } from "./transport.js";
import { argsProblem, lookupParams } from "./validate.js";

/** What is wrong with a value, and WHERE. The path is Playwright's format —
 * `a.inner[1].property` — because naming the offending path is the difference
 * between a caller fixing the bug in a minute and bisecting their own arguments. */
export interface WireProblem {
  /** Dotted/bracketed path from the root. Empty string at the root. */
  path: string;
  /** What was found and why the wire refuses it, in one clause. */
  reason: string;
}

const PLAIN_PROTOTYPES: ReadonlyArray<object | null> = [Object.prototype, null];

/** Walk `value` and return the FIRST thing the wire refuses, or `null` when it is
 * clean.
 *
 * `undefined` is allowed, at the root and at a key — it is what an absent optional
 * property is, and refusing it would refuse every legal partial object. It is
 * DROPPED on the way across, which `Wire<T>` states in the type. */
export function wireProblem(value: unknown): WireProblem | null {
  return walk(value, "", new Set<object>());
}

function walk(value: unknown, path: string, seen: Set<object>): WireProblem | null {
  const bad = (reason: string): WireProblem => ({ path, reason });

  if (value === null || value === undefined) return null;

  switch (typeof value) {
    case "string":
    case "boolean":
      return null;
    case "number":
      // JSON.stringify turns these into `null` without complaint, which is the
      // silent-lossy class this guard exists to catch.
      return Number.isFinite(value) ? null : bad(`the number ${String(value)} has no JSON form`);
    case "bigint":
      return bad("a bigint has no JSON form (JSON.stringify throws on one)");
    case "function":
      return bad("a function cannot cross the wire — code never crosses, only data");
    case "symbol":
      return bad("a symbol has no JSON form");
  }

  const object = value as object;
  if (seen.has(object)) return bad("a circular reference");
  seen.add(object);
  try {
    if (Array.isArray(object)) {
      for (let i = 0; i < object.length; i++) {
        const problem = walk(object[i], `${path}[${i}]`, seen);
        if (problem) return problem;
      }
      return null;
    }

    // A NOMINAL refusal, and the only one available at runtime. `Wire<T>` cannot
    // express it — TypeScript has no nominal typing — so a Date, a Map, a class
    // instance and anything else carrying its own prototype is refused here rather
    // than being quietly flattened into a plain object or a string.
    const prototype = Object.getPrototypeOf(object) as object | null;
    if (!PLAIN_PROTOTYPES.includes(prototype)) {
      return bad(`a ${constructorName(object)} instance — only plain objects and arrays cross`);
    }

    for (const key of Object.keys(object)) {
      const child = (object as Record<string, unknown>)[key];
      const childPath = path ? `${path}.${key}` : key;
      // A function- or symbol-valued key is DROPPED rather than refused: that is
      // what JSON.stringify does, and `Wire<T>` removes the key from its key set to
      // say so. Refusing here would make a legal object with a method unusable.
      if (typeof child === "function" || typeof child === "symbol") continue;
      const problem = walk(child, childPath, seen);
      if (problem) return problem;
    }
    return null;
  } finally {
    seen.delete(object);
  }
}

function constructorName(object: object): string {
  const name = (Object.getPrototypeOf(object) as { constructor?: { name?: string } } | null)
    ?.constructor?.name;
  return name && name.length > 0 ? name : "non-plain-object";
}

/** Refuse the whole argument list before a byte leaves the process.
 *
 * Throws `BowmarkError` with code `wire_refused`, naming the exact position:
 * `bowmark.hotels.search(args[0].checkIn)`. Failure is loud here for the reason EF
 * Core made silent client-side fallback an error: a boundary that quietly accepts
 * something it will mangle produces behaviour that only breaks in production, and
 * getting better at serialization later must not silently change a caller's
 * results. */
export function assertWireSafeArgs(label: string, args: readonly unknown[]): void {
  for (let i = 0; i < args.length; i++) {
    const problem = wireProblem(args[i]);
    if (!problem) continue;
    const where = problem.path
      ? problem.path.startsWith("[")
        ? `args[${i}]${problem.path}`
        : `args[${i}].${problem.path}`
      : `args[${i}]`;
    throw new BowmarkError(
      `${label} was not called: ${where} is ${problem.reason}. ` +
        "Only JSON — plain objects, arrays, strings, finite numbers, booleans and null — crosses to a capability.",
      { code: "wire_refused", path: label },
    );
  }
}

/** Refuse an argument list the declared signature does not accept, and refuse a path
 * this package has never heard of.
 *
 * Runs AFTER `assertWireSafeArgs`, deliberately. A `Date` and a `Map` are refused by
 * the wire guard with a message about JSON, which is the right explanation; reaching
 * the shape check first would report the same value as "expected a string" and send
 * the caller looking for the wrong bug.
 *
 * ── FAILING CLOSED, and the two things it must NOT close on ─────────────────
 *
 * The rule is "refuse what the table knows is wrong; pass anything it cannot know",
 * and both halves cost something real, so both are stated.
 *
 * **A known unit with an unknown FUNCTION is refused.** The table is authoritative
 * about a unit it carries, so `bowmark.music.searchHarder(…)` is a typo or an
 * install older than the function. The compile-time surface already refuses that
 * call — it is a `Property does not exist` error against the same generated data —
 * so anybody reaching this line came through `as any`, plain JavaScript or a stale
 * package, and a refusal naming the manifest version is a better answer for all
 * three than a request that succeeds against declarations they do not have. The
 * cost, named rather than hidden: this package is published on its own cadence, so
 * a caller on version N cannot reach a function the library gained in N+1 even
 * though the api would serve it. `run(script)` reaches anything and is untyped by
 * construction, so nothing is unreachable.
 *
 * **An unknown UNIT passes straight through, and that is not a hedge.** A Shopify
 * family MEMBER — `bowmark.providers.gymshark.search(…)` — is deliberately absent
 * from every manifest: `listProviders()` excludes members by design and always
 * will, because there are half a million of them. So an unknown unit is the NORMAL
 * case for most of the library, not a stale one, and refusing it would have this
 * package refuse the largest part of what it is a client for. Found by running the
 * Phase 5 session suite, whose worked example is gymshark.
 *
 * **An `unchecked` function passes through too** — one whose declared argument is a
 * bare destructuring pattern. It is an EXPLICIT null in the table rather than an
 * absence, and that distinction is what makes the first rule safe at all: 20 real
 * functions have no readable argument shape, and refusing them would delete them
 * from the runtime as well as from the types. */
export function assertArgShape(
  label: string,
  path: readonly string[],
  args: readonly unknown[],
): void {
  const found = lookupParams(VALIDATORS, path);
  if (found.kind === "unknown-function") {
    throw new BowmarkError(
      `${label} was not called: this package's declarations were generated from library ` +
        `manifest ${VALIDATORS.version.slice(0, 12)}, which has no such function on that unit. ` +
        `If it is newer than this package, upgrade @bowmark/web; if you meant a different name, ` +
        `the typed surface will offer it. \`run(script)\` reaches anything, typed or not.`,
      { code: "unknown_function", path: label },
    );
  }
  if (found.kind === "unknown-unit" || found.kind === "unchecked") return;

  const unit = VALIDATORS.units[path.slice(0, -1).join(".")];
  const problem = argsProblem(found.params, args, unit?.defs ?? {});
  if (!problem) return;
  throw new BowmarkError(`${label} was not called: ${problem.path} is ${problem.reason}.`, {
    code: "bad_argument",
    path: label,
  });
}
